pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -157,6 +157,16 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
 		}
 	}
 
+	// cfs quota for container is not capped if
+	// 1. cpu manager policy is static
+	// 2. pod has quos PodQOSGuaranteed
+	// 3. container has integer cpu request
+	if m.containerManager.cpuExperimentalManagerPolicyStatic &&
+		kubeapiqos.GetPodQOS(pod) == v1.PodQOSGuaranteed &&
+		container.Resources.Requests.Cpu().MilliValue() % MilliCPUToCPU == 0 {
```

Comment:
this logic should really be in the cpumanager. In https://github.com/kubernetes/kubernetes/pull/117030 we check the container got exclusive cpu assigned; we can redo the math instead, there are pros/cons for both approaches, but we should not replicate the logic outside cpumanager.

Reply:
totally agree would like to see the logic in a single place. unfortunately there is no single shared method I see between sandbox config, pod resource config, and container resource configs.

# Pod Resources

LinuxContainerResources Quota for pod resources is computed in `cm.ResourceConfigForPod`. This method is called from several places, so the computation can't be moved "up" without duplicating the logic, and it can't be moved "down" further to another method because it's built in this method directly using the pod resources. 

```
cm.ResourceConfigForPod
  (constructs ResourceConfig directly from pod resources)
``

calls
* kubelet.SyncPod -> containermanager.EnsureExists -> containermanager.ResourceConfigForPod
* pkg/volume/emptydir/empty_dir
* kuberuntime.SyncPod -> kuberuntime.doPodResizeAction -> containermanager.ResourceConfigForPod

# Sandbox and Container Resources

Right now these are computed, both resources are fetched from `calculateLinuxResources` and then the quota is overwritten when the pinning qualifications are met.

```
calculateSandboxResources -> calculateLinuxResources
generateLinuxContainerConfig -> generateLinuxContainerResources -> calculateLinuxResources
```

# Single-Location Fix? 

The smoothest fix would be to modify `calculateLinuxResources`, that takes care of both container and sandbox computations. Then we could modify ResourceConfigForPod to use `calculateLinuxResources` instead of computing the values directly. The single method would then be used in all cases. could also lock that additional computation behind a feature gate

Reply:
I'm thinking about adding the functionality to query if a pod got (or will get) exclusive CPU assignment in cpumanager, and calling cpumanager in the relevant flows. https://github.com/kubernetes/kubernetes/pull/117030 already does something like it.

Reply:
`pkg/volume/emptydir/empty_dir` seems to be using `ResourceConfigForPod` only to compute in a consistent way the memory values. In this case we can just hardcoded to false the exclusive CPU assignment flag, with a note explaining why.